ports/stm32/uart.c

```diff
@@ -536,6 +541,15 @@ void uart_irq_handler(mp_uint_t uart_id) {
             }
         }
     }
+    if (__HAL_UART_GET_FLAG(&self->uart, UART_FLAG_IDLE) != RESET) {
+        __HAL_UART_CLEAR_IDLEFLAG(&self->uart);
+        self->irq.flags |= IRQ_FLAG_RX_IDLE;
+    }
+    self->irq.flags &= self->irq.enable;
+    if (self->irq.flags != IRQ_FLAG_NONE) {
+        irq_dispatch(&self->irq);
```

Comment:
I think there should be 2 changes here: 1) pass `self` as the second arg which should be then passed to the callback, to eliminate the need for the `parent` entry in the irq object; 2) if the function returns an error then disable the IRQ (like timer.c does it).